crates/consensus/beacon/src/engine/hooks/mod.rs

```diff
@@ -0,0 +1,116 @@
+use reth_interfaces::sync::SyncState;
+use reth_primitives::BlockNumber;
+use std::task::{Context, Poll};
+
+mod controller;
+pub(crate) use controller::HooksController;
+
+mod prune;
+pub use prune::PruneHook;
+
+/// Collection of [hooks][`Hook`].
+#[derive(Default)]
+pub struct Hooks {
+    inner: Vec<Box<dyn Hook>>,
+}
+
+impl Hooks {
+    /// Creates a new empty collection of [hooks][`Hook`].
+    pub fn new() -> Self {
+        Self { inner: Vec::new() }
+    }
+
+    /// Adds a new [hook][`Hook`] to the collection.
+    pub fn add<H: Hook>(&mut self, hook: H) {
+        self.inner.push(Box::new(hook))
+    }
+}
+
+/// Hook that will be run during the main loop of
+/// [consensus engine][`crate::engine::BeaconConsensusEngine`].
+pub trait Hook: Send + Sync + 'static {
+    /// Returns a human-readable name for the hook.
+    fn name(&self) -> &'static str;
+
+    /// Advances the hook execution, emitting an [event][`HookEvent`].
+    fn poll(&mut self, cx: &mut Context<'_>, args: HookArguments) -> Poll<HookEvent>;
+
+    /// Returns an [action][`HookAction`], if any, according to the passed [event][`HookEvent`].
+    fn on_event(&mut self, event: HookEvent) -> Result<Option<HookAction>, HookError>;
+
+    /// Returns [dependencies][`HookDependencies`] for running this hook.
+    fn dependencies(&self) -> HookDependencies;
```

Comment:
this implies the dependencies are required for the duration of the hook which might not be the case.

I think we can solve this a bit differently, for example with multiple poll functions:

like `poll_start -> Poll<Deps>`, 
and `poll_hook()`

Reply:
not following... 

> this implies the dependencies are required for the duration of the hook which might not be the case.

for DB write access, this is the case. Not sure what other dependencies we might have in the future though, so you might be right, but should we be ready for this now?

Reply:
simplified according to https://github.com/paradigmxyz/reth/pull/4582#discussion_r1326878819